### ISSUE TITLE
[Taskbar height and icon size] Taskbar pushed off screen and gap below full screen apps

### ISSUE THREAD

**Issue by placoderm** [ramensoftware/windhawk-mods#859]
Mod version: 1.2.10
Windhawk: 1.5.1
VivoBook ASUS touchscreen
Windows 11, 23H2 KB5040527
OS Builds 22621.3958 and 22631.3958
[log.txt](https://github.com/user-attachments/files/16606241/log.txt)

Note: prior to installing Windhawk I had been on the "get latest updates as soon as they are available." I have since turned that off.

There are two problems
1. The task bar is "pushed" down off the screen and items are not centered vertically
2. Apps that are maximized to full screen keep a gap between them and the task bar.

I have tried changing the screen scale and display resolution and neither has changed anything

I have tried hibernating as well as rebooting. This does not fix the problem.

Here is what my taskbar normally looks like:
![image](https://github.com/user-attachments/assets/2d57fe25-6d72-4a2f-a0b6-fa119c4527bb)

And here is what it looks like after activating:

![2024-08-14 10_57_06-screenshot](https://github.com/user-attachments/assets/c93ae4da-5ba8-4b1e-afeb-0ed2c6c74254)
Note the "ghost" start button and odd thing in the middle. This disappears when I restart the machine.

Making the icon height much smaller does not fix either problem:
![2024-08-14 11_12_14-screenshot](https://github.com/user-attachments/assets/546ea469-e256-4a17-9a71-3fa3a0f21691)

And if I make the height small enough, the taskbar can go off the screen completely:
![2024-08-14 11_13_00-screenshot](https://github.com/user-attachments/assets/1cad9fdf-f950-419d-ba94-27c691c1eb19)

FYI @m417z 

It seems related to but different from https://github.com/ramensoftware/windhawk-mods/issues/568 and https://github.com/ramensoftware/windhawk-mods/issues/529. Seems very similar to https://github.com/ramensoftware/windhawk-mods/issues/413 but without the off center problem.

BTW, Thanks so much for this software. It's wonderful. I'm looking forward to getting this mod to work.

**Comment by RagibHasin** [comment]
As I understand the issues, it isn't different from #529. This is the same issue I'm facing. And your report confirms that it only occurs on devices with a touch screen.

**Comment by placoderm** [comment]
> it isn't different from #529.

Well, to be clear it is different in the sense that rebooting and restarting explorer doesn't fix it (If I am reading your issue correctly. Also, It doesn't confirm that it is _only_ on touch screen. But for sure it occurs on mine 😁

I wanted to mention that I'm happy to do any testing or provide any further information.

**Comment by RagibHasin** [comment]
Restarting explorer used to fix this problem on an earlier update, it does not now. And you are right, possibility of this appearing on non-touchscreen device is not ruled out 😉.

**Comment by placoderm** [comment]
> it does not now.

Oh! Apologies. @m417z if this is really a double then feel free to close.

**Comment by jmain33** [comment]
Add me to the pack... I'm on a Surface Tablet with the same precise experience.

Also the Copilot icon seems to end up pushed way down, with the entire notification/control center section out of alignment; which may be what's offsetting the entire taskbar.

![Screenshot 2024-08-14 111144](https://github.com/user-attachments/assets/12673797-2c64-421c-971e-c50f7f332bb8)

So, maybe, this is a more unique problem than the older tickets.

**Comment by crystalfunky** [comment]
I don't know where my thread went but I reported this problem some days ago as well. So hopefully this will be fixed soon.

**Comment by ziyi-yan** [comment]
I've encountered the same issue and my machine also has a touchscreen.

**Comment by ziyi-yan** [comment]
Another observation I have about this issue is that it only happens on the main display. When I attach an external monitor, the task bar on that monitor is not pushed off screen with this mod enabled but the task bar on my laptop monitor is still off screen.
When I close the lid of my laptop, the task bar on the external monitor becomes off screen and I suspect that it's because that external monitor is showing the main display.

**Comment by disperaller** [comment]
+1

**Comment by RagibHasin** [comment]
I have been digging through the code trying to understand it, and have some questions for @ziyi-yan.
1. Does your external monitor also have touch capabilities?
2. What happens when you shut your laptop lid and restart explorer on the external monitor? Does the taskbar remain shifted or is it back to normal?

**Comment by Splinter047** [comment]
Suffering from the same issue, again laptop with a touch screen, a lenovo yoga 6 in my case. restarting explorer does nothing to fix it, disabling only fixes after you restart explorer.

**Comment by m417z** [comment]
I think that #529 is different - more users are reporting this issue, and with #529, restarting explorer worked as a workaround.

I wasn't able to reproduce this issue. Can anyone try version 1.2.9 to check whether the issue was introduced in version 1.2.10?

To install 1.2.9, disable the existing mod, then click the bottom right button to create a new mod, replace the code with [the content from here](https://github.com/ramensoftware/windhawk-mods/blob/7e0bd27ae1d12ae639497fbc9b48bb791f98b078/mods/taskbar-icon-size.wh.cpp), click "Compile Mod", then "Exit Editing Mode".

**Comment by RagibHasin** [comment]
Yes! Thank you, @m417z! v1.2.9 works for me as intended.

**Comment by jmain33** [comment]
I did as requested and can report that v1.2.9 runs without causing this issue, the same as @RagibHasin says.
Even the issue #726 which I myself reported a couple of months ago seems  to get repaired with version 1.2.9?

![Screenshot 2024-08-16 151727](https://github.com/user-attachments/assets/f72f65ab-b56e-4ae1-91d3-f3e0b1a8b723)

**Comment by ziyi-yan** [comment]
Hi @RagibHasin,
> 1. Does your external monitor also have touch capabilities?
No, it's not a touchscreen.
> 2. What happens when you shut your laptop lid and restart explorer on the external monitor? Does the taskbar remain shifted or is it back to normal?
No, it doesn't go back normal after I restart explorer.

**Comment by placoderm** [comment]
> Can anyone try version 1.2.9

I can confirm that this works after restarting explorer.exe. It took some fiddling with the size numbers to get everything in the right place, but eventually it looks correct. And as a reminder, I'm using a laptop with touch screen.

Funny thing is, I've been using the large (default) icons for so long now I'm not sure I even like the smaller ones 🤣.

Thank you @m417z for all of your work. I'm happy to keep testing things if you need.

**Comment by ziyi-yan** [comment]
@m417z I've installed v1.2.9 according to your instructions. It was initially off screen but became normal after I restarted explorer.exe.

**Comment by m417z** [comment]
Thanks for confirming. I couldn't reproduce it on my computer, VM, or laptop with a touch screen, so I need your help. Please try these versions and let me know which ones are working correctly and which ones have the issue. It will help understand which of the changes between 1.2.9 and 1.2.10 causes the problem:
[taskbar-icon-size-test.zip](https://github.com/user-attachments/files/16641758/taskbar-icon-size-test.zip)

**Comment by placoderm** [comment]
> Please try these versions

Sorry… So you would like us to repeat the steps you gave us [here](https://github.com/ramensoftware/windhawk-mods/issues/859#issuecomment-2293870017) for each one of these and tell you what works and what doesn't work? I'm assuming since they all have identical names wee need to remove the old one before we start the new one.

**Comment by jmain33** [comment]
@ziyi-yan, does the taskbar for the external monitor get altered by Windhawk at all?
You stated that it does not get offset by v1.2.10 of this mod… but does it properly get the effects of your Windhawk mods applied, or is it just showing an unaltered taskbar?

I'm not a developer, just curious.

**Comment by m417z** [comment]
@placoderm yes. There's no need to remove them, you can just replace the content. The steps can be similar to the following:

* Disable the existing mod.
* Click the bottom right button to create a new mod.
* Replace the code with the content of taskbar-icon-size-test1.wh.cpp.
* Click "Compile Mod".
* Click on "Preview Mod" and change the mod settings as needed.
* If needed, restart explorer or reboot, and see if there's a problem.

Then for the other files:

* Replace the code with the content of taskbar-icon-size-test2.wh.cpp.
* Click "Compile Mod".
* If needed, restart explorer or reboot, and see if there's a problem.
* Replace the code with the content of taskbar-icon-size-test3.wh.cpp.
* ...

**Comment by jmain33** [comment]
**Note for all testers:** it is actually necessary to restart Windows Explorer after compiling each test file, as Michael says, or you will not see the true effects from the mod file.

@m417z **:** test files 1-3 each looked completely as they should be expected, test files 4-6 all offset the taskbar in the same way as one another. Additionally; all the test files show the weather widget as it should look, even the ones that offset the taskbar, while v1.2.10 still causes it to have a blank graphic displayed. So it looks as though #726 might be truthfully fixed.

**Comment by placoderm** [comment]
> Please try these versions

Works:
* test1
* test2
* test3

Doesn't work (pushed down, max windows gap):

* test4
* test5
* test6

I should also add that restarting `explorer.exe` in the task manager *didn't* always instigate the change (working or not). The only reliable method was to restart it by changing the icon width setting and then saving the setting, which pops up the dialog to restart explorer.

**Comment by crystalfunky** [comment]
> > Please try these versions
> 
> Works:
> 
>     * test1
> 
>     * test2
> 
>     * test3
> 
> 
> Doesn't work (pushed down, max windows gap):
> 
>     * test4
> 
>     * test5
> 
>     * test6
> 
> 
> I should also add that restarting `explorer.exe` in the task manager _didn't_ always instigate the change (working or not). The only reliable method was to restart it by changing the icon width setting and then saving the setting, which pops up the dialog to restart explorer.

None of that worked for me unfortunately

**Comment by Splinter047** [comment]
so 1.2.9 works after restarting explorer, however there is another issue where when you fold the laptop for tablet mode, the taskbar is supposed to shrink and a bar appears to slide it up, with the mod tho, the space becomes blank and unfolding the laptop glitches the taskbar and windows until you restart explorer.
Going to check out the tests you shared next.

**Comment by m417z** [comment]
Thanks, now I know what change causes the issue, although I don't quite understand why it's necessary. I restored it, please confirm that this version works correctly:
https://github.com/m417z/my-windhawk-mods/blob/3fb58a8e130750046a99b8f33e2ad3f482ea5bdf/mods/taskbar-icon-size.wh.cpp

**Comment by jmain33** [comment]
@m417z - thank you, this seems OK for me.

**Comment by placoderm** [comment]
> please confirm that this version works correctly:

Works for me! Thanks.

**Comment by RagibHasin** [comment]
Works for me too! Thank you.

**Comment by crystalfunky** [comment]
> Thanks, now I know what change causes the issue, although I don't quite understand why it's necessary. I restored it, please confirm that this version works correctly: https://github.com/m417z/my-windhawk-mods/blob/3fb58a8e130750046a99b8f33e2ad3f482ea5bdf/mods/taskbar-icon-size.wh.cpp

That's working for me! For now.

**Comment by m417z** [comment]
Thank you for testing, I released version 1.2.11 with the fix.

One of the concerns with the fix is that it can no longer be used in the next Windows version (Win11 24H2), that's why I removed this code in 1.2.10. To find the best alternative solution, I need to understand why it's necessary to begin with. Doing it without having a way to reproduce it is tricky, so I'll going to ask you guys to run versions with extra logs and post the log output.

Let's start with this one, please post the logs that it prints:
https://gist.github.com/m417z/d772003b49d3df27efefaf77e63eb48c

* Disable the existing mod.
* Click the bottom right button to create a new mod.
* Replace the code with the content of the code in the link.
* Turn on "Enable logging".
* Click on "Show Log Output".
* Click "Compile Mod".
* You should see logs printed at the bottom, wait for 5-10 seconds, then copy the output and post it here.

**Comment by placoderm** [comment]
OK, after I clicked compile, here is the log: [placoderm-log.txt](https://github.com/user-attachments/files/16651932/placoderm-log.txt)
Works as expected.

**Comment by jmain33** [comment]
When it was done compiling the taskbar was offset (this is likely a result of the older bug where that can initially happen), I then restarted Windows Explorer which resulted in a normally displayed taskbar. The logging below captured all of this:
[taskbar-height-and-icon-size-1.2.11-test1.txt](https://github.com/user-attachments/files/16652050/taskbar-height-and-icon-size-1.2.11-test1.txt)

**Comment by m417z** [comment]
OK, thanks for the logs, it's a bit clearer now. The alternative hook I used in 1.2.10 is called at a later stage, apparently too late. Unfortunately, I didn't find a better alternative for Win11 24H2, not a simple one at least.

I'm going to close this issue for now as it seems to be fixed for everybody with 1.2.11, but the bug might still be there for Win11 24H2. I can't check it as I don't have a device with the issue, but keep an eye on it and let me know if you update and it returns.

**Comment by jmain33** [comment]
Thank you @m417z, I'll report back if anything arises in the coming Windows update.

**Comment by placoderm** [comment]
> I'm going to close this issue for now

Thank you!!!

**Comment by jmain33** [comment]
If any of the other users of the "Taskbar Height and Icon Size" mod are still paying attention to this... Since you are all also users of devices with touch screens, could you please try increasing the icon size of to a value higher than 32px to see if it causes the taskbar's weather widget to go blank? And report in thread #726 if you encounter the same problem.

@m417z has an inclination that issue #726 for this same mod may also have to do with touch capable devices.
Thank you

**Comment by Splinter047** [comment]
Still having the same issue of it glitching out when in folded mode, its supposed to turn into a bar quite similar to the iphone gesture bar but instead it disappears below the screen, and when you try to extend it by swiping everything is glitched, using 1.2.12 on windows 23h2.

**Comment by m417z** [comment]
@Splinter047 tablet mode isn't supported, see https://github.com/ramensoftware/windhawk-mods/issues/184.
